### PR TITLE
Implement Quadlog

### PR DESCRIPTION
**Pull request by vikusc**

### CODE REVIEW

**Comment by mreboud** [comment]
That's weird, the bin file is not found on `jammy` but everything work on MacOS...
I will test locally on my laptop.

**Comment by mreboud** [comment]
Found it I think, @vikusc can you add your new file to the EXTRA_DIST in the Makefile.am? Here:
https://github.com/vikusc/eos/blob/c730628d988d9e0b087a34673e27f4624b47b017/eos/maths/Makefile.am#L31

File: eos/maths/Makefile.am [hunk 1]
```diff
@@ -26,6 +26,7 @@ libeosmaths_la_CXXFLAGS = $(AM_CXXFLAGS) \
 libeosmaths_la_LDFLAGS = $(AM_LDFLAGS) $(GSL_LDFLAGS)
 
 EXTRA_DIST = \
+	polylog_TEST_quadlog.bin \
```
Comment (mreboud) [inline eos/maths/Makefile.am:29]: Can we keep alphabetic order here?

File: eos/maths/polylog.cc [hunk 2]
```diff
@@ -15,6 +33,7 @@ namespace eos
 
     namespace dilog_impl
     {
+        // for the dilogarithm the coefficients of f1 are : Zeta(2 - m) / m! starting from m = 0 to m = max_iterations
```
Comment (mreboud) [inline eos/maths/polylog.cc:36]: Why is this comment here and not above `f1`?

File: eos/maths/polylog.cc [hunk 3]
```diff
@@ -243,12 +277,137 @@ namespace eos
...
+
+            return result;
+        }
+
+        // used for |z| > 2.0
+        complex<double> g(const complex<double> & z)
+        {
+            const complex<double> lnz = std::log(z);
+            complex<double> a = - 2.0 / 3.0 * power_of<4>(M_PI);
+            complex<double> b = complex<double>(0.0, -0.5 / M_PI) * lnz;
```
Comment (mreboud) [inline eos/maths/polylog.cc:377]: `a` and `b` can be made `const`, right?

File: eos/maths/polylog.cc [hunk 4]
```diff
@@ -243,12 +277,137 @@ namespace eos
...
+
+            return result;
+        }
+
+        // used for |z| >= 0.5 and |z| <= 2.0
+        complex<double> f1(const complex<double> & z)
+        {
+            complex<double> result(0.0, 0.0);
+
+            complex<double> lnz = std::log(z);
```
Comment (mreboud) [inline eos/maths/polylog.cc:350]: `lnz` can be made `const`, correct?

File: eos/maths/polylog.cc [hunk 5]
```diff
@@ -243,12 +277,137 @@ namespace eos
...
+
+    // Calculation of the quadlogarithm based on [C:2006A]
+    complex<double> quadlog(const complex<double> & z)
+    {
+        // special cases
+        if (z == complex<double>(0.0, 0.0))
+            return complex<double>(0.0);
+
+        if (z == complex<double>(1.0, 0.0))
+            return M_PI * M_PI * M_PI * M_PI / 90.0;
```
Comment (mreboud) [inline eos/maths/polylog.cc:400]: Can you use `power_of<4>`?

File: eos/maths/polylog.cc [hunk 6]
```diff
@@ -243,12 +277,137 @@ namespace eos
...
+
+        if (z == complex<double>(-1.0, 0.0))
+            return - M_PI * M_PI * M_PI * M_PI * 7.0 / 720.0;
```
Comment (mreboud) [inline eos/maths/polylog.cc:403]: Idem

File: eos/maths/polylog.hh [hunk 7]
```diff
@@ -10,6 +28,9 @@ namespace eos
     complex<double> dilog(const complex<double> & z) __attribute__ ((pure));
 
     complex<double> trilog(const complex<double> & z) __attribute__ ((pure));
+
+    complex<double> quadlog(const complex<double> & z) __attribute__ ((pure));
+
 }
 
-#endif
+#endif
```
Comment (mreboud) [inline eos/maths/polylog.hh:36]: Can you add a newline at the end of the file?

**Review by mreboud — CHANGES_REQUESTED** [review]
I only have a couple of minor comments, the rest looks good!

File: eos/maths/polylog.cc (hunk 2 above)
Comment (vikusc) [reply eos/maths/polylog.cc:36]: the list in the line below contains the series coefficients of f1

File: eos/maths/Makefile.am (hunk 1 above)
Comment (vikusc) [reply eos/maths/Makefile.am:29]: Yes, I did not notice the ordering.

File: eos/maths/polylog.cc (hunk 2 above)
Comment (dvandyk) [reply eos/maths/polylog.cc:36]: I agree with @vikusc here, the comment is well placed.

**Review by mreboud — APPROVED** [review]

**Pull request merged**